src/Storages/MergeTree/MergeTreeSequentialSource.cpp

```diff
@@ -214,8 +216,32 @@ try
 
     for (size_t i = 0; i < result_header.columns(); ++i)
     {
-        auto pos = reader_header.getPositionByName(result_header.safeGetByPosition(i).name);
+        const auto & name = result_header.safeGetByPosition(i).name;
+        auto pos = reader_header.getPositionByName(name);
         auto & result_column = result_columns.emplace_back(std::move(read_result.columns[pos]));
+        if (name == "_part_offset" && read_task_info->merged_part_offsets)
+        {
+            if (read_task_info->merged_part_offsets->isFinalized())
+            {
+                /// Projection part
+
+                result_column = result_column->convertToFullColumnIfSparse();
+                auto & column = result_column->assumeMutableRef();
+                auto & offset_data = assert_cast<ColumnUInt64 &>(column).getData();
+                for (auto & offset : offset_data)
+                    offset = (*read_task_info->merged_part_offsets)[offset + read_task_info->starting_offset];
+            }
+            else if (read_task_info->starting_offset > 0)
+            {
+                /// Parent part
+
+                result_column = result_column->convertToFullColumnIfSparse();
+                auto & column = result_column->assumeMutableRef();
+                auto & offset_data = assert_cast<ColumnUInt64 &>(column).getData();
+                for (auto & offset : offset_data)
+                    offset += read_task_info->starting_offset;
+            }
+        }
```

Comment:
This part looks good, but can we add a comment about what we do with the `_part_offset` value here?

Also, it might be a bit confusing that for the projection part  `_part_offset` is a position in the source part, but not inside the projection. Maybe we can have a different name for it? E.g. `_parent_part_offset` or `_source_part_offset`. (They can be visible and different for `mergeTreeProjection` table function, but automatically renamed `_parent_part_offset -> _part_offset` on select if the projection was used).

Reply:
> Also, it might be a bit confusing that for the projection part `_part_offset` is a position in the source part, but not inside the projection

Actually, I feel the opposite. In normal projections, all columns are derived from `SELECT .. FROM <source_part>`. So `SELECT _part_offset` being based on the source part feels consistent with the rest. Also, it wouldn't really make sense to support a `_part_offset` that refers to a position within the projection part itself...

Reply:
> If we store some null value in mapping, it could mean that the row was replaced, and we can filter it out.

Yeah, it's possible. But it would require deliberate design of the FOR encoding to distinguish and handle those null values correctly.

Reply:
I see.

Reply:
> It makes sense for mergeTreeProjection function; in the same way as it makes sense for the normal table: I can use it for debugging projection. If one day _parent_part_offset is incorrectly written, I could (potentially) figure out in which row in the projection part it was.

For this purpose, maybe we can introduce `_projection_part_offset` column instead? It's just a bit weird to do that `_part_offset` -> `_parent_part_offset` mapping around projection query analysis.

> When I read the random part of MergeTask.cpp, I don't understand right away what _part_offset means.

If that only affects code reading, I'll try to add more comments. TBH, this makes implementation easier and cleaner...

Reply:
To me it looks much clearer to have `_part_offset` as a non-physical virtual column for all data parts (no matter parent part of projection part) and `_source_part_offset` as a physical virtual column only for projection parts. Naming is not important, but it's clear enough, that these two columns are significantly different.
`_source_part_offset`:
1. May exist only in project part
2. Stored on disk, i.e. physical
3. Points to the row in a **different data part**.
`_part_offset`:
1. Exists in both projection parts and parent parts
2. Not stored on disk, pure virtual
3. Just a number of row in this data part.

Reply:
Ok. Let's spare a dedicated name for this column, i.e. if a table is created with column `_parent_part_offset` or physical `_part_offset`, we will not allow creating projections with `_part_offset`.